src/lib.rs

```diff
@@ -1965,3 +1953,21 @@ fn connect(addr: SockAddr, domain: Domain, protocol: Option<Protocol>) -> io::Re
     }
     Ok(socket)
 }
+
+#[derive(Debug)]
+struct ArcSource(Arc<Source>);
+
+impl Drop for ArcSource {
+    fn drop(&mut self) {
+        // Deregister and ignore errors because destructors should not panic.
+        Reactor::get().remove_io(&self.0).ok();
+    }
+}
+
+impl Deref for ArcSource {
+    type Target = Source;
+
+    fn deref(&self) -> &Self::Target {
+        self.0.deref()
+    }
+}
```

Comment:
I'd rather have this in `reactor.rs` and named something a little better, like `SourceContainer` or something.

Reply:
done